packages/next/src/server/app-render/app-render.tsx

```diff
@@ -1777,38 +1790,589 @@ async function prerenderToStream(
+        // is if it completes in a microtask and that's likely very rare for any non-trivial app
+        ;(
+          prerenderAsyncStorage.run(
+            // The store to scope
+            prospectiveRenderPrerenderStore,
+            // The function to run
+            ComponentMod.prerender,
+            // ... the arguments for the function to run
+            firstAttemptRSCPayload,
+            clientReferenceManifest.clientModules,
+            {
+              nonce: ctx.nonce,
+              onError,
+              signal: flightController.signal,
+            }
+          ) as Promise<ReactServerPrerenderResolveToType>
+        ).catch(() => {})
+
+        // When this resolves the cache has no inflight reads and we can ascertain the dynamic outcome
+        await cacheSignal.cacheReady()
+        if (reactServerIsDynamic) {
+          // During a prospective render the only dynamic thing that can happen is a synchronous dynamic
+          // API access. We expect to have a tracked expression to use for our dynamic error but we fall back
+          // to a generic error if we don't.
+          const dynamicReason = getFirstDynamicReason(dynamicTracking)
+          if (dynamicReason) {
+            throw new DynamicServerError(
+              `Route ${staticGenerationStore.route} couldn't be rendered statically because it used \`${dynamicReason}\`. See more info here: https://nextjs.org/docs/messages/dynamic-server-error`
+            )
+          } else {
+            console.error(
+              'Expected Next.js to keep track of reason for opting out of static rendering but one was not found. This is a bug in Next.js'
+            )
+            throw new DynamicServerError(
+              `Route ${staticGenerationStore.route} couldn't be rendered statically because it used a dynamic API. See more info here: https://nextjs.org/docs/messages/dynamic-server-error`
+            )
+          }
+        } else {
+          // The render didn't explicitly use any Dynamic APIs but it might have IO so we need to retry
+          // the render. We abort the current render here to avoid doing unecessary work.
+          // Keep in mind that while the render is aborted, inflight async ServerComponents can still continue
+          // and might call dynamic APIs.
+          flightController.abort(abortReason)
+        }
+
+        // Reset the prerenderState because we are going to retry the render
+        flightController = new AbortController()
+        dynamicTracking = createDynamicTrackingState(
+          renderOpts.isDebugDynamicAccesses
+        )
+        reactServerIsDynamic = false
+
+        const finalRenderPrerenderStore: PrerenderStore = {
+          // During the final prerender we don't need to track cache access so we omit the signal
+          cacheSignal: null,
+          controller: flightController,
+          dynamicTracking,
+        }
+
+        const finalAttemptRSCPayload = await getRSCPayload(
```

Comment:
looks like this rsc payload and reactServerResult generation are similar between static and postpone case, maybe we can consider sharing it in the later refactors. it's quite long to go through the whole process.

Reply:
Yeah I initially tried but there is a conflict with metadata resolution that required a new tree to be created. Definitely want to circle back to it though